File: examples/java/src/main/java/ai/starlake/quack/example/Query.java
package ai.starlake.quack.example;

/**
 * Run a single SQL statement against the Quack-on-Demand FlightSQL edge and print the result.
 *
 * <pre>
 *   mvn -q compile exec:java
 *   mvn -q compile exec:java -Dexec.args="SELECT count(*) FROM tpch1.customer"
 * </pre>
 *
 * See {@link QodClient} for the JDBC plumbing. Connection settings come from {@code QOD_*} env vars
 * (see README).
 */
public final class Query {

  private static final String DEMO =
      "SELECT * FROM (VALUES (1, 'duck'), (2, 'quack'), (3, 'lake')) AS t(id, label)";

  public static void main(String[] args) {
    String sql = args.length > 0 ? args[0] : envOrDefault("QOD_SQL", DEMO);
    try (QodClient client = QodClient.connect(QodConfig.fromEnv())) {
      System.out.println("Connecting to " + client.describe());
      System.out.println("SQL: " + sql + "\n");

      QueryResult result = client.query(sql);

      System.out.println("columns: " + result.columnSummary());
      System.out.println("rows: " + result.rowCount() + "\n");
      int limit = Math.min(result.rowCount(), 100);
      for (int i = 0; i < limit; i++) {
        System.out.println(result.rowAsJson(i));
      }
    } catch (Exception err) {
      System.err.println("\nquery failed: " + err.getMessage());
      System.exit(1);
    }
  }

  private static String envOrDefault(String name, String fallback) {
    String value = System.getenv(name);
    return value == null || value.isEmpty() ? fallback : value;
  }
}

File: examples/rust/src/bin/query.rs
//! Run a single SQL statement against the Quack-on-Demand FlightSQL edge and
//! print the Arrow result.
//!
//! Run:  cargo run --bin query
//!       cargo run --bin query -- "SELECT count(*) FROM tpch1.customer"
//!
//! See `src/lib.rs` for the tonic + Arrow plumbing. Connection settings come
//! from QOD_* env vars (see README).

use qod::{QodClient, QodConfig};

const DEMO: &str = "SELECT * FROM (VALUES (1, 'duck'), (2, 'quack'), (3, 'lake')) AS t(id, label)";

#[tokio::main]
async fn main() {
    if let Err(err) = run().await {
        eprint!("\nquery failed: {err}");
        let mut source = err.source();
        while let Some(cause) = source {
            eprint!("\n  caused by: {cause}");
            source = cause.source();
        }
        eprintln!();
        std::process::exit(1);
    }
}

async fn run() -> Result<(), Box<dyn std::error::Error>> {
    let sql = std::env::args()
        .nth(1)
        .or_else(|| std::env::var("QOD_SQL").ok())
        .unwrap_or_else(|| DEMO.to_string());

    let mut client = QodClient::connect(QodConfig::from_env()).await?;
    println!("Connecting to {}", client.describe());
    println!("SQL: {sql}\n");

    let result = client.query(&sql).await?;

    println!("columns: {}", result.column_summary());
    println!("rows: {}\n", result.num_rows());
    let limit = result.num_rows().min(100);
    for i in 0..limit {
        println!("{}", result.row_as_json(i)?);
    }

    Ok(())
}
